text/0000-cosign.md

```diff
@@ -0,0 +1,109 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+As a part of the exporting process, the `analyzer` and `exporter` will now take an addition flag `-cosign-config <cosign.toml>` which will contain the cosign configuration.
+
+The schema for `cosign.toml` will look like the following - 
+
+```toml
+# A static list of annotations to add to each signature config
+[[annotations]]
+name = "<name>"
+value = "<value>"
+
+[[annotations]]
+name = "<name2>"
+value = "<value2>"
+
+# multiple cosign configs can be specified
+[[configs]]
+# path to private key
+key = "<path-to-key>"
+# Exclusive OR with `key`
+# private key inline as a b64 encoded string
+key-data = "<base64-encoded-private-key>"
+# optional value for COSIGN_REPOSITORY, optional, default:""
+repository = "<cosign-repository>"
+# value for COSIGN_DOCKER_MEDIA_TYPES, optional, default:false
+docker-media-types = true
+# optional path to password file. Contents will be used as COSIGN_PASSWORD
+password = "<path-to-password-file>"
+# Exclusive OR with `password`
+# optional base64 encoded value for COSIGN_PASSWORD
+password-data = "<base64-encoded-password>"
+# optional flag to export sbom if present, default:true
+export-sbom = true
+# optional flag to sign sbom if exported, default: true
+sign-sbom = true
+```
+
+The `analyzer` will be responsible for verifying registry access to push the required signatures. The exporter will be responsible for generating and pushing the signatures defined by each of the configs above for each tag specified to the `exporter`. As a part of this process, the lifecycle will also strip out any environment variables that start with `COSIGN_` before analyzing or exporting cosign specific tasks.
+
+During the export phase, the lifecycle will add all the provided annotations to each exported signature. It will also add the corresponding image tag and build time of the image as tags to the signature.
+The lifecycle will also export and sign related SBOM artifacts from the build. The format for attaching SBOM to the image will follow the specification described at https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md.
+
+The above will only be supported if `lifecycle` is run in `registry` mode. If we wish to support it in local mode, we may add support for this in case we drop daemon mode and introdue an OCI layout mode.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The exporter can use `cosign` as a library to generate, attach and publish signatures for the output image. `cosign` exposes a GGCR compatible interface which should fit nicely with the lifecycle.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Increased complexity in lifecycle
```

Comment:
I recall we discussed this in sub sync, but is it possible to add the advantage of having the lifecycle do the signing vs leaving it up to the platform?

Reply:
How does it increase the complexity here ? Could someone add more details ?

Reply:
If this moves outside `lifecycle` to being a platform responsibility, does the binary belong within the lifecycle distribution?

Reply:
Probably not... but in this case `pack create builder` should be expanded to include optional platform-provided binaries within the builder at the desired location.